docs/source/user-guide/concepts/expressions-and-contexts.md

```diff
@@ -0,0 +1,192 @@
+
+{{code_block('user-guide/concepts/expressions','filter-1',['filter'])}}
+
+```python exec="on" result="text" session="user-guide/concepts/expressions-and-contexts"
+--8<-- "python/user-guide/concepts/expressions.py:filter-1"
+```
+
+### `group_by` and aggregations
+
+In the context `group_by`, rows are grouped according to the unique values of the grouping expressions.
+You can then apply expressions to the resulting groups, which may be of variable lengths.
+
+When using the context `group_by`, you can use an expression to compute the groupings dynamically:
+
+{{code_block('user-guide/concepts/expressions','group_by-1',['group_by'])}}
+
+```python exec="on" result="text" session="user-guide/concepts/expressions-and-contexts"
+--8<-- "python/user-guide/concepts/expressions.py:group_by-1"
+```
+
+After using `group_by` we use `agg` to apply aggregating expressions to the groups.
+Since in the example above we only specified the name of a column, we get the groups of that column as lists.
+
+We can specify as many grouping expressions as we'd like and the context `group_by` will group the rows according to the distinct values across the expressions specified.
+Here, we group by a combination of decade of birth and whether the person is shorter than 1.7 metres:
+
+{{code_block('user-guide/concepts/expressions','group_by-2',['group_by'])}}
+
+```python exec="on" result="text" session="user-guide/concepts/expressions-and-contexts"
+--8<-- "python/user-guide/concepts/expressions.py:group_by-2"
+```
+
+The resulting dataframe, after applying aggregating expressions, contains one column per each grouping expression on the left and then as many columns as needed to represent the results of the aggregating expressions.
+In turn, we can specify as many aggregating expressions as we want:
+
+{{code_block('user-guide/concepts/expressions','group_by-3',['group_by'])}}
+
+```python exec="on" result="text" session="user-guide/concepts/expressions-and-contexts"
+--8<-- "python/user-guide/concepts/expressions.py:group_by-3"
+```
+
+See also `group_by_dynamic` and `group_by_rolling` for other grouping contexts.
+
+## Expression expansion
+
+The last example contained two grouping expressions and three aggregating expressions, and yet the resulting dataframe contained six columns instead of five.
+If we look closely, the last aggregating expression mentioned two different columns: “weight” and “height”.
+
+Polars expressions support a feature called _expression expansion_.
+Expression expansion is like a shorthand notation for when you want to apply the same transform to multiple columns.
+As we have seen, the expression
+
+```python
+pl.col("weight", "height").mean().name.prefix("avg_")
+```
+
+will compute the mean value of the columns “weight” and “height” and will rename them as “avg_weight” and “avg_height”, respectively.
+
+In this case, this expression expands into two independent expressions that Polars can execute in parallel.
```

Comment:
Maybe to make it extra clear, print a list with the expanded expressions? I think holding hands can help people with their mental model.

Reply:
No, I meant by hand. Just to help them understand what happens under the hood if expressions "expand".